pkg/kptfile/kptfileutil/util.go

```diff
@@ -182,3 +172,280 @@ func HasKptfile(path string) (bool, error) {
+// and updateDef, skip copy the local node if nothing changed or updateDef get deleted.
+// It deletes the node from updateDef if node get deleted in localDef
+func mergeDef(updatedDef, localDef, originalDef *yaml.MapNode) error {
+	localKeys, err := localDef.Value.Fields()
+	if err != nil {
+		return err
+	}
+	updatedKeys, err := updatedDef.Value.Fields()
+	if err != nil {
+		return nil
+	}
+	keys := append(updatedKeys, localKeys...)
+
+	unique := make(map[string]bool, len(keys))
+	for _, key := range keys {
+		if unique[key] {
+			continue
+		}
+		unique[key] = true
+
+		node := localDef.Value.Field(key)
+		if node == nil {
+			node = updatedDef.Value.Field(key)
+		}
+
+		if shouldSkipCopy(updatedDef, localDef, originalDef, key) {
+			continue
+		}
+
+		if shouldRemoveValue(updatedDef, localDef, originalDef, key) {
+			err = updatedDef.Value.PipeE(yaml.FieldClearer{Name: key})
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		err = updatedDef.Value.PipeE(yaml.FieldSetter{
+			Name:  key,
+			Value: node.Value})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// shouldSkipCopy decides if a node with key should be copied from fromDef to toDef
+func shouldSkipCopy(updatedDef, localDef, originalDef *yaml.MapNode, key string) bool {
+	if originalDef == nil || updatedDef == nil || localDef == nil {
+		return false
+	}
+	localVal := localDef.Value.Field(key)
+	originalVal := originalDef.Value.Field(key)
+	updatedVal := updatedDef.Value.Field(key)
+	if localVal == nil || originalVal == nil {
+		return false
+	}
+
+	localValStr, err := localVal.Value.String()
+	if err != nil {
+		return false
+	}
+	originalValStr, err := originalVal.Value.String()
+	if err != nil {
+		return false
+	}
+
+	// skip copying if the definition is deleted from upstream
+	if updatedVal == nil {
+		return true
+	}
+	// skip copying if original val matches with from val(local val)
+	return localValStr == originalValStr
+}
+
+// shouldRemoveValue decides if a node with key should be removed from Def
+func shouldRemoveValue(updatedDef, localDef, originalDef *yaml.MapNode, key string) bool {
+	localVal := localDef.Value.Field(key)
+	originalVal := originalDef.Value.Field(key)
+	updatedVal := updatedDef.Value.Field(key)
+
+	if originalVal == nil || updatedVal == nil {
+		return false
+	}
+
+	originalValStr, err := originalVal.Value.String()
+	if err != nil {
+		return false
+	}
+
+	updatedValStr, err := updatedVal.Value.String()
+	if err != nil {
+		return false
+	}
+
+	if localVal == nil && originalValStr == updatedValStr {
+		return true
+	}
+
+	return false
+}
+
+// MergeSubpackages takes the subpackage information from local, updated
+// and original and does a 3-way merge. The result is returned as a new slice.
+// The passed in data structures are not changed.
+func MergeSubpackages(local, updated, original []v1alpha2.Subpackage) ([]v1alpha2.Subpackage, error) {
```

Comment:
Why does this show up as all new? I think there are changes to this code that isn't captured here.

Reply:
Because it is moved from pkgfile.go to util.go. pkgfile.go hosted the types previously and now they are moved to types.go. So felt it is more cleaner to move them here and delete pkgfile.go and pkgfile_test.go

Reply:
But the code in the deleted function (https://github.com/GoogleContainerTools/kpt/pull/1479/files#diff-8ba9bcc40fc6e1016b348f9534a95c77e15fbb3aea9ecfcdb37e3c928c4a3c40L268) is not the same as the code here. I assume things got mixed up in a rebase, but verify that we aren't loosing updates.

Reply:
Yes. Moved and rebased. The changes are captured and tests are passing.